public/client/TracyProfiler.cpp

```diff
@@ -1484,22 +1502,30 @@ void Profiler::InstallCrashHandler()
     sigaction( SIGPIPE, &crashHandler, &m_prevSignal.pipe );
     sigaction( SIGBUS, &crashHandler, &m_prevSignal.bus );
     sigaction( SIGABRT, &crashHandler, &m_prevSignal.abrt );
-#endif
 
-#if defined _WIN32 && !defined TRACY_UWP && !defined TRACY_NO_CRASH_HANDLER
-    m_exceptionHandler = AddVectoredExceptionHandler( 1, CrashFilter );
+    m_crashHandlerInstalled = true;
 #endif
 
-#ifndef TRACY_NO_CRASH_HANDLER
-    m_crashHandlerInstalled = true;
+#if defined _WIN32 && !defined TRACY_UWP && !defined TRACY_NO_CRASH_HANDLER
+    // We cannot use Vectored Exception handling because it catches application-wide frame-based SEH blocks. We only
+    // want to catch unhandled exceptions in the event that there is not already an unhandled exception filter.
+    if( auto prev = SetUnhandledExceptionFilter( CrashFilter ) )
+        SetUnhandledExceptionFilter( prev ); // Already had a handler => put it back
```

Comment:
What is the reason for this change?

Reply:
Vectored exception handlers are called for every exception, including frame-based exception handlers that handle them. So any Windows application that would use Tracy along with frame-based SEH blocks would inadvertently cause `CrashFilter` to be called.

Instead, we want `CrashFilter` to be called only for unhandled exceptions, hence the change to `SetUnhandledExceptionFilter`.

The answer on [this page](https://stackoverflow.com/questions/54806802/filtering-critical-fatal-exceptions-for-use-with-addvectoredexceptionhandler/54879453#54879453) gives more information, specifically this portion:
> VEH handlers added by AddVectoredExceptionHandler will handle exceptions before they reach frame-based handlers.
> 
> Filter set by SetUnhandledExceptionFilter will handle exceptions after frame-based handlers failed.

Reply:
Btw, here's a simple program that exemplifies what I'm talking about with vectored exceptions. If Tracy is connected, `CrashFilter` is called when the `memcpy` causes an access violation. If Tracy is not connected, the `CrashFilter` isn't installed and the program runs fine.
```cpp
int main(int argc, char** argv)
{
    tracy::StartupProfiler();

    std::this_thread::sleep_for(std::chrono::seconds(1)); // give time to connect

    {
        ZoneNamed(temp, true);

        __try
        {
            memcpy(nullptr, &temp, 1); // should crash?
        }
        __except (1)
        {
            // nope, control should end up here instead preventing the crash
        }

        std::this_thread::sleep_for(std::chrono::milliseconds(10));
    }

    tracy::ShutdownProfiler();

    return 0;   
}
```

Reply:
Sigh. The `memcpy` above is ub, so any reasoning about what should happen is nonsensical. After #887 I don't have any energy left for Microsoft bullshit right now.

Reply:
Yes, this simple example is UB, but the point is to exemplify a scenario where any Structured Exception is triggered and instead of being handled, the `CrashFilter` is invoked instead.

And yeah, understandable--887 looks rough.

Reply:
Alright, here's another example. Integer divide-by-zero can be intercepted and handled with structured exceptions. If this happens while Tracy is connected with its vectored exception handler however, the profiler/viewer reports a crash and the application hangs. If Tracy is not connected, the vectored exception handler is not installed, and the application prints "Divide-by-zero without crash" as the exception was handled.
```cpp
        __try
        {
            int i = 1;
            int j = 1;
            --j;
            fprintf(stdout, "%d\n", i / j);
        }
        __except (1)
        {
            fprintf(stdout, "Divide-by-zero without crash\n");
        }
```

This of course merely illustrates that using vectored exception handlers will override _any_ frame-based exception handlers in an undesirable fashion, when what we really want is to handle _unhandled_ exceptions.

Yet another example is the old method of setting thread names for Visual Studio:
```cpp
        THREADNAME_INFO info = {};
        info.dwType = 0x1000;
        info.szName = name;
        info.dwThreadID = threadId;
        __try
        {
            RaiseException(MS_VC_EXCEPTION, 0, sizeof(info) / sizeof(ULONG_PTR), (ULONG_PTR*)&info);
        }
        __except (EXCEPTION_EXECUTE_HANDLER)
        {
        }
```
this too inadvertently results in Tracy's `CrashFilter` being invoked as currently written, however in this case it is not one of the recognized exceptions and returns `EXCEPTION_CONTINUE_SEARCH` which allows the frame-based filter to be called.

In all of these cases, my change to use `SetUnhandledExceptionFilter` causes `CrashFilter` to be invoked only in the event of an actual unhandled exception.

Reply:
Thanks for the explanation, it does make sense.

My initial question was actually for the "Already had a handler => put it back" bit, which seems to deny enabling the crash handler, when something else is already hooked up. The `TRACY_NO_CRASH_HANDLER` macro is intended to handle this case.

Reply:
I pushed a change that makes Windows now always set the unhandled exception filter, similar to how Linux always sets the signal handlers.

In a modern modular world the way these handlers work is not great, but such is the OS design that we have to work with. The fact that we have `TRACY_NO_CRASH_HANDLER` gives a work-around which is good.

I also changed the Linux side so that it matches the Windows removal-side now: if something else overrode our signal/SEH handler we put it back.